File: Frontend/src/app/service/premio.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Premio } from '../Models/Premio';

@Injectable({
  providedIn: 'root',
})
export class PremioService {
  private apiUrl = 'http://localhost:8080/premios';

  constructor(private http: HttpClient) {}

  // Obtener todos los premios
  findAll(): Observable<Premio[]> {
    return this.http.get<Premio[]>(this.apiUrl);
  }

  // Obtener premios disponibles 
  getPremiosDisponibles(): Observable<Premio[]> {
    return this.http.get<Premio[]>(`${this.apiUrl}/disponibles`);
  }

  // Obtener premio por ID
  findById(id: number): Observable<Premio> {
    return this.http.get<Premio>(`${this.apiUrl}/${id}`);
  }

  // Obtener premios ganados por un usuario
  getPremiosByUsuario(usuarioId: number): Observable<Premio[]> {
    return this.http.get<Premio[]>(`${this.apiUrl}/usuario/${usuarioId}`);
  }

  // Obtener premios de una reserva
  getPremiosByReserva(reservaId: number): Observable<Premio[]> {
    return this.http.get<Premio[]>(`${this.apiUrl}/reserva/${reservaId}`);
  }

  // Asignar premio a usuario y reserva
  asignarPremio(premioId: number, usuarioId: number, reservaId: number): Observable<Premio> {
    const payload = {
      premioId: premioId,
      usuarioId: usuarioId,
      reservaId: reservaId
    };
    return this.http.post<Premio>(`${this.apiUrl}/asignar`, payload);
  }

  // Crear nuevo premio 
  create(premio: Premio): Observable<Premio> {
    return this.http.post<Premio>(this.apiUrl, premio);
  }

  // Eliminar premio por ID
  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: Backend/src/main/java/com/example/lumaresort/service/PremioService.java
package com.example.lumaresort.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.example.lumaresort.entities.Premio;
import com.example.lumaresort.entities.Reserva;
import com.example.lumaresort.entities.Usuario;
import com.example.lumaresort.repository.PremioRepository;
import com.example.lumaresort.repository.ReservaRepository;
import com.example.lumaresort.repository.UsuarioRepository;

@Service
public class PremioService {

    @Autowired
    private PremioRepository premioRepository;

    @Autowired
    private UsuarioRepository usuarioRepository;

    @Autowired
    private ReservaRepository reservaRepository;

    public List<Premio> findAll() {
        return premioRepository.findAll();
    }

    public Premio save(Premio premio) {
        return premioRepository.save(premio);
    }

    public void delete(Premio premio) {
        premioRepository.delete(premio);
    }

    public Premio findById(Long id) {
        return premioRepository.findById(id).orElse(null);
    }

    // Obtener premios disponibles
    public List<Premio> getPremiosDisponibles() {
        return premioRepository.findByUsuarioIsNull();
    }

    // Obtener premios ganados por un usuario
    public List<Premio> getPremiosByUsuario(Long usuarioId) {
        return premioRepository.findByUsuarioIdUsuario(usuarioId);
    }

    // Obtener premios de una reserva
    public List<Premio> getPremiosByReserva(Long reservaId) {
        return premioRepository.findByReservaIdReserva(reservaId);
    }

    // Asignar premio a un usuario y reserva
    public Premio asignarPremio(Long premioPlantillaId, Long usuarioId, Long reservaId) {
        Premio premioPlantilla = premioRepository.findById(premioPlantillaId).orElse(null);
        if (premioPlantilla == null) {
            return null;
        }

        // Buscar usuario y reserva
        Usuario usuario = usuarioRepository.findById(usuarioId).orElse(null);
        Reserva reserva = reservaRepository.findById(reservaId).orElse(null);

        if (usuario == null || reserva == null) {
            return null;
        }

        // Crear una nueva instancia del premio y asignar al usuario y reserva
        Premio premioAsignado = Premio.builder()
                .nombre(premioPlantilla.getNombre())
                .descripcion(premioPlantilla.getDescripcion())
                .color(premioPlantilla.getColor())
                .icono(premioPlantilla.getIcono())
                .usuario(usuario)
                .reserva(reserva)
                .build();

        return premioRepository.save(premioAsignado);
    }
}
